File: core/node/events/migrations/snapshot_migration_0001_test.go
package migrations

import (
	"testing"

	"github.com/river-build/river/core/node/base/test"
	"github.com/river-build/river/core/node/crypto"
	. "github.com/river-build/river/core/node/protocol"
	"github.com/river-build/river/core/node/testutils"
	"github.com/stretchr/testify/require"
)

// nasty bug with the insert_sorted function, it was inserting an extra element at the end
// every insert, we need to remove duplicates

func TestSnapshotMigration0001(t *testing.T) {
	ctx, cancel := test.NewTestContext()
	defer cancel()
	userWallet, _ := crypto.NewWallet(ctx)
	spaceId := testutils.FakeStreamId(0x10) // events.STREAM_SPACE_BIN
	channelId := testutils.MakeChannelId(spaceId)

	// snaps have multiple member instances
	badMemberSnap := &Snapshot{
		Members: &MemberPayload_Snapshot{
			Joined: []*MemberPayload_Snapshot_Member{
				{
					UserAddress: userWallet.Address[:],
				},
				{
					UserAddress: userWallet.Address[:],
				},
			},
		},
	}
	// migrate
	migratedSnapshot := snapshot_migration_0001(badMemberSnap)
	require.Equal(t, 1, len(migratedSnapshot.Members.Joined))

	// space channel payloads
	badSpaceChannel := &Snapshot{
		Content: &Snapshot_SpaceContent{
			SpaceContent: &SpacePayload_Snapshot{
				Channels: []*SpacePayload_ChannelMetadata{
					{
						ChannelId: channelId[:],
					},
					{
						ChannelId: channelId[:],
					},
				},
			},
		},
	}
	migratedSnapshot = snapshot_migration_0001(badSpaceChannel)
	require.Equal(t, 1, len(migratedSnapshot.GetSpaceContent().Channels))

	// user payload user membership
	badUserPayload := &Snapshot{
		Content: &Snapshot_UserContent{
			UserContent: &UserPayload_Snapshot{
				Memberships: []*UserPayload_UserMembership{
					{
						StreamId: spaceId[:],
					},
					{
						StreamId: spaceId[:],
					},
				},
			},
		},
	}
	migratedSnapshot = snapshot_migration_0001(badUserPayload)
	require.Equal(t, 1, len(migratedSnapshot.GetUserContent().Memberships))

	// user settings fully read markers
	badUserSettings := &Snapshot{
		Content: &Snapshot_UserSettingsContent{
			UserSettingsContent: &UserSettingsPayload_Snapshot{
				FullyReadMarkers: []*UserSettingsPayload_FullyReadMarkers{
					{
						StreamId: channelId[:],
					},
					{
						StreamId: channelId[:],
					},
				},
			},
		},
	}
	migratedSnapshot = snapshot_migration_0001(badUserSettings)
	require.Equal(t, 1, len(migratedSnapshot.GetUserSettingsContent().FullyReadMarkers))
}

File: packages/sdk/src/migrations/snapshotMigration0001.test.ts
/**
 * @group main
 */

import { Snapshot } from '@river-build/proto'
import { snapshotMigration0001 } from './snapshotMigration0001'
import { ethers } from 'ethers'
import { makeUniqueSpaceStreamId } from '../util.test'
import { addressFromUserId, makeUniqueChannelStreamId, streamIdAsBytes } from '../id'
import { check } from '@river-build/dlog'

// a no-op migration test for the initial snapshot, use as a template for new migrations
describe('snapshotMigration0001', () => {
    test('run migration', () => {
        const wallet = ethers.Wallet.createRandom()
        const userAddress = addressFromUserId(wallet.address)
        const spaceIdStr = makeUniqueSpaceStreamId()
        const spaceId = streamIdAsBytes(spaceIdStr)
        const channelIdStr = makeUniqueChannelStreamId(spaceIdStr)
        const channelId = streamIdAsBytes(channelIdStr)

        // members
        const badMemberSnap = new Snapshot({
            members: {
                joined: [{ userAddress: userAddress }, { userAddress: userAddress }],
            },
        })
        const result = snapshotMigration0001(badMemberSnap)
        expect(result.members?.joined.length).toBe(1)

        // space channel payloads
        const badSpaceSnap = new Snapshot({
            content: {
                case: 'spaceContent',
                value: {
                    channels: [{ channelId: channelId }, { channelId: channelId }],
                },
            },
        })
        const result2 = snapshotMigration0001(badSpaceSnap)
        check(result2.content.case === 'spaceContent')
        expect(result2.content?.value.channels.length).toBe(1)

        // user payload
        const badUserPayload = new Snapshot({
            content: {
                case: 'userContent',
                value: {
                    memberships: [{ streamId: spaceId }, { streamId: spaceId }],
                },
            },
        })
        const result3 = snapshotMigration0001(badUserPayload)
        check(result3.content.case === 'userContent')
        expect(result3.content?.value.memberships.length).toBe(1)

        // user settings
        const badUserSettings = new Snapshot({
            content: {
                case: 'userSettingsContent',
                value: {
                    fullyReadMarkers: [{ streamId: spaceId }, { streamId: spaceId }],
                },
            },
        })
        const result4 = snapshotMigration0001(badUserSettings)
        check(result4.content.case === 'userSettingsContent')
        expect(result4.content?.value.fullyReadMarkers.length).toBe(1)
    })
})
